### PR TITLE
h3: earlier check for control and QPACK stream termination

### PR DESCRIPTION
**Pull request by LPardue**
h3: earlier checks for control and QPACK stream termination

A peer can FIN or reset a control stream at any point during the stream
lifetime. While doing so is an error in HTTP/3, our handling before this change
was a bit odd.

Previously, depending on the sequencing, an unexpected termination of such a
stream could trigger an InvalidStreamState or StreamReset error during poll().
This would bubble up to the app, which is weird and unexpected because the
quiche h3 layer is supposed to deal with all the control stream details. While
both those errors contain a stream ID, the app isn't given any info that the ID
is a control stream.

While any subsequent call to poll() would then cause the stream_finished
detection to kick in and enforce the RFC rules to close the connection with
CloseCriticialStream, that's a bit late and clunky.

This change adds additional checks for the control and QPACK stream terminations
in order to make sure we close connections at the earliest time and prevent
useless errors getting bubbled up to apps.

### CODE REVIEW

File: quiche/src/h3/mod.rs [hunk 1]
```diff
@@ -931,6 +931,26 @@ pub struct Stats {
...
+    conn.close(
+        true,
+        Error::ClosedCriticalStream.to_wire(),
+        b"Critical stream closed.",
+    )?;
+
+    Err(Error::ClosedCriticalStream)
+}
+
+fn check_closed_critical_stream(
```
Comment (ghedo) [inline quiche/src/h3/mod.rs:944]:
```suggestion
fn close_finished_critical_stream_with_error(
```

Bit of a mouthful, but seems more accurate?

Comment (ghedo) [reply quiche/src/h3/mod.rs:944]: Or, idk, `close_critical_stream_if_finished(`

Comment (LPardue) [reply quiche/src/h3/mod.rs]: I went with `close_conn_if_critical_stream_finished()`

Comment (ghedo) [reply quiche/src/h3/mod.rs]: Oh lol, yeah, that makes way more sense.

**Review by ghedo — APPROVED** [review]

**Pull request merged**